### PR TITLE
configurable wgpu backend

### PR DESCRIPTION
**Pull request by Wumpf**
introduces `WgpuConfiguration`, a config object used for both Native (winit) and Web rendering with wgpu. Encapsulates creation options for wgpu instance/adapter/device/surface and exposes a callback for surface errors.

Needed for users selecting WebGPU as backend on the web - #2137 

(tested so far only via demo on wgpu and wgpu/web(gl))

### CODE REVIEW

File: crates/eframe/CHANGELOG.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ NOTE: [`egui-winit`](../egui-winit/CHANGELOG.md), [`egui_glium`](../egui_glium/C
 * Added `center` to `NativeOptions` and `monitor_size` to `WindowInfo` on desktop ([#2035](https://github.com/emilk/egui/pull/2035)).
 * Web: you can access your application from JS using `AppRunner::app_mut`. See `crates/egui_demo_app/src/lib.rs`.
 * Web: You can now use WebGL on top of `wgpu` by enabling the `wgpu` feature (and disabling `glow` via disabling default features) ([#2107](https://github.com/emilk/egui/pull/2107)).
-
+* Wgpu device/adapter/surface creation has now various configuration options exposed via `NativeOptions/WebOptions::wgpu_options`
```
Comment (emilk) [inline crates/eframe/CHANGELOG.md:17]: link to PR!

File: crates/eframe/src/epi.rs [hunk 2]
```diff
@@ -467,8 +480,25 @@ impl Default for WebOptions {
...
+
+            #[cfg(feature = "wgpu")]
+            wgpu_options: egui_wgpu::WgpuConfiguration {
+                // WebGPU is not stable enough yet, use WebGL emulation
+                backends: wgpu::Backends::GL,
+                device_descriptor: wgpu::DeviceDescriptor {
+                    label: Some("egui wgpu device"),
+                    features: wgpu::Features::default(),
+                    limits: wgpu::Limits {
+                        // When using a depth buffer, we have to be able to create a texture large enough for the entire surface.
```
Comment (emilk) [inline crates/eframe/src/epi.rs:495]:
```suggestion
                        // When using a depth buffer, we have to be able to create a texture
                        // large enough for the entire surface, and we want to support 4k+ displays.
```

**Review by emilk — CHANGES_REQUESTED** [review]
This is great ⭐

File: crates/egui-wgpu/src/lib.rs [hunk 3]
```diff
@@ -30,6 +29,60 @@ pub struct RenderState {
     pub renderer: Arc<RwLock<Renderer>>,
 }
 
+/// Specifies which action should be taken as consequence of a [`wgpu::SurfaceError`]
+pub enum SurfaceErrorAction {
+    /// Do nothing and skip the current frame.
+    SkipFrame,
+
+    /// Instructs egui to recreate the surface.
```
Comment (teh-cmc) [inline crates/egui-wgpu/src/lib.rs:37]:
```suggestion
/// Instructs egui to recreate the surface, then skip the current frame.
```

**Review by teh-cmc — APPROVED** [review]

File: crates/egui-wgpu/CHANGELOG.md [hunk 4]
```diff
@@ -11,6 +11,7 @@ All notable changes to the `egui-wgpu` integration will be noted in this file.
 * `PrepareCallback` now passes `wgpu::CommandEncoder` ([#2136](https://github.com/emilk/egui/pull/2136))
 * Only a single vertex & index buffer is now created and resized when necessary (previously, vertex/index buffers were allocated for every mesh) ([#2148](https://github.com/emilk/egui/pull/2148)).
 * `Renderer::update_texture` no longer creates a new `wgpu::Sampler` with every new texture ([#2198](https://github.com/emilk/egui/pull/2198))
+* `Painter`'s device/adapater/surface creation is now configurable via `WgpuConfiguration`
```
[the remark this answers is not in the document]
Comment (Mingun) [reply crates/egui-wgpu/CHANGELOG.md:14]: Also misprint "adapater" -> "adapter"

**Review by emilk — APPROVED** [review]

**Pull request merged**